### DISCUSSION TITLE
Sanchez scripting

### DISCUSSION THREAD

**Discussion by parsoli** [JVital2013/vitality-goes#35] [General]
Hey there,

So yesterday I ran Sanchez and she ran for hours.  Created a bunch of images and all of those images look great!  At the end of the run of images, I received the below errors.  The Sanchez log looks clean, but consists of only processing the images themselves.

xplanet was installed and can be found at:
./etc/xplanet
./usr/bin/xplanet
./usr/share/doc/xplanet
./usr/share/xplanet

in .\scripts\Resources, I do have all of the jpg's and 12 files called xplanet-01 - 12

In .\scripts\sanchez-v1.0.22-linux-x64\ I have a couple of JSON files and 3 world.x.jpg files

Wiped all files out of \cache, no impact

in Scriptconfig.ini I only have (1) Source path and (1) Destination path
sanchezSrcPath16=/var/www/html/NASLink/goes16/fd/ch13
sanchezDstPath16=/var/www/html/NASLink/goes16/fd/sanchez
and one Sanchez path
sanchezPath=/home/sparsons/vitality-goes/scripts/sanchez-v1.0.22-linux-x64/Sanchez

Any thoughts on why this would throw these errors?  I'm unsure the operation it's trying to accomplish after it renders the last jpg (which it does complete).

sparsons@goesview:~/vitality-goes/scripts$ sudo ./Sanchez.sh
[2022-12-22 13:42:12] Creating GOES 16 Sanchez False Color Images
[2022-12-22 13:42:24] Creating .DS_Store...
date: invalid date ‘Store St:or: + 5 hour’
Warning: Can't find /home/sparsons/vitality-goes/scripts/Resources/xplanet- in
xplanet/config
/root/.xplanet/config
/usr/share/xplanet/config
**Error: Can't load configuration file /home/sparsons/vitality-goes/scripts/Resources/xplanet-**
**Exiting from readConfig.cpp at line 580
identify-im6.q16: no decode delegate for this image format `' @ error/constitute.c/ReadImage/575.**
**./Sanchez.sh: line 105: 10850/: syntax error: operand expected (error token is "/")**
[2022-12-22 13:42:24] Creating GOES 17 Sanchez False Color Images
find: ‘’: No such file or directory
[2022-12-22 13:42:24] Creating GOES 18 Sanchez False Color Images
find: ‘’: No such file or directory
[2022-12-22 13:42:24] Creating GOES 16/17 Composite False Color Images
find: ‘’: No such file or directory
[2022-12-22 13:42:24] Creating GOES 16/18 Composite False Color Images
find: ‘’: No such file or directory

**Comment by JVital2013** [comment]
I'm guessing you used a Mac to view the images directly on disk at some point? It looks like there's a stray `.DS_STORE` file in your `sanchezSrcPath16` path. It would be a hidden file created by macOS.

Right now the Sanchez script throws errors if there's a stray file in the directory. You can ignore the error, or delete the hidden `.DS_STORE` file from within your `ch13` folder.

Those `find: ‘’: No such file or directory` errors are normal for GOES-17 and 18 since you don't have directories specified and can be ignored.

**Reply by parsoli** [reply]
Ah, well I do use a MAC as well to monitor the file systems.  I piped out my Cron job for Sanchez.  Will any of this be an issue?  xplanet is in path's and can be launched from anywhere from what I can tell.  Also another question.  Sanchez only processed my FC CH13 files up until Dec 13th.  I wonder if that's what the .DS_STORE caused.

sparsons@goesview:~$ cat cronsanchez.log
/home/sparsons/vitality-goes/scripts/Sanchez.sh: line 2: readlink: command not found
/home/sparsons/vitality-goes/scripts/Sanchez.sh: line 2: dirname: command not found
/home/sparsons/vitality-goes/scripts/Sanchez.sh: line 2: /scriptconfig.ini: No such file or directory
xplanet could not be found, which is required for this script

**Reply by JVital2013** [reply]
The 13th cutoff could be because of the .DS_STORE file. Shut off any finder-based mac monitoring for now, delete the .DS_STORE file, and try again. If it goes through fine, the .DS_STORE file is to blame.

**I will update Sanchez.sh to exclude extraneous files like this - be on the lookout for it!**

Onto the problems you have when running it through cron: can you send me the line you're using for cron to invoke the script? It almost looks like `sh` is trying to run the script instead of `bash`. The script needs to be run by `bash`

**Reply by parsoli** [reply]
Man, since all my files are on a Synology, had to do some creative things to first find the .DS_Store files, but then banish write of those in the future.  Thanks for that tip!  Now Sanchez is processing all the rest of the images.

For the Cron, if I need bash, I need to add "#!/bin/sh" to the top of the crontab right?  Today, the detail is below.  The data above the detail is all # out Ubuntu stuff with no "#!/bin/sh" at the top.

PATH=/bin/sh:/home/sparsons/vitality-goes/scripts
0 0 * * *  /home/sparsons/vitality-goes/scripts/CreateVideos-ABI.sh > /home/sparsons/croncreatevideosabi.log 2>&1
0 2 * * *  /home/sparsons/vitality-goes/scripts/CreateVideos-EMWIN.sh > /home/sparsons/croncreatevideosemwin.log 2>&1
55 11 * * * /home/sparsons/vitality-goes/scripts/Cleanup-EmwinText.sh > /home/sparsons/cronemwincleanup.log 2>&1
25,55 2-23 * * * /home/sparsons/vitality-goes/scripts/Sanchez.sh > /home/sparsons/cronsanchez.log 2>&1

**Reply by JVital2013** [reply]
Bash requires `#!/bin/bash` at the top of the script. `#!/bin/sh` will launch the sh shell instead. https://www.geeksforgeeks.org/difference-between-sh-and-bash/ looks to have a good overview of the difference. Sanchez.sh should have `#!/bin/bash` by default.

That being said, crontab uses sh by default. I believe you can put `SHELL=/bin/bash` at the top of your crontab to force it to use bash. https://unix.stackexchange.com/questions/94456/how-to-change-cron-shell-sh-to-bash talks a bit about it.

Something like this may work (untested)
```
SHELL=/bin/bash
0 0 * * * /home/sparsons/vitality-goes/scripts/CreateVideos-ABI.sh > /home/sparsons/croncreatevideosabi.log 2>&1
0 2 * * * /home/sparsons/vitality-goes/scripts/CreateVideos-EMWIN.sh > /home/sparsons/croncreatevideosemwin.log 2>&1
55 11 * * * /home/sparsons/vitality-goes/scripts/Cleanup-EmwinText.sh > /home/sparsons/cronemwincleanup.log 2>&1
25,55 2-23 * * * /home/sparsons/vitality-goes/scripts/Sanchez.sh > /home/sparsons/cronsanchez.log 2>&1
```

**Reply by parsoli** [reply]
No dice overnight.  It's an environment path issue I think.  All 4 scripts tried to run.  Of course, if I run them individually with sudo, they work fine or else I'd have other problems.  Crontab log shows:

CreatevideoABI received :  mogrify could not be found, which is required for this script
CreatevideoEMWIN received: ffmpeg could not be found, which is required for this script
Emwincleanup received:  zip could not be found, which is required for this script
Sanchez receives:
/home/sparsons/vitality-goes/scripts/Sanchez.sh: line 2: readlink: command not found
/home/sparsons/vitality-goes/scripts/Sanchez.sh: line 2: dirname: command not found
/home/sparsons/vitality-goes/scripts/Sanchez.sh: line 2: /scriptconfig.ini: No such file or directory
xplanet could not be found, which is required for this script

Last night I made the PATH and SHELL adjustments shown below.  Didn't help at all

PATH=/bin/sh:/usr/bin/zip:/usr/bin/xplanet:/home/sparsons/vitality-goes/scripts
SHELL=/bin/bash
0 0 * * *  /home/sparsons/vitality-goes/scripts/CreateVideos-ABI.sh > /home/sparsons/croncreatevideosabi.log 2>&1
0 2 * * *  /home/sparsons/vitality-goes/scripts/CreateVideos-EMWIN.sh > /home/sparsons/croncreatevideosemwin.log 2>&1
55 11 * * * /home/sparsons/vitality-goes/scripts/Cleanup-EmwinText.sh > /home/sparsons/cronemwincleanup.log 2>&1
25,55 2-23 * * * /home/sparsons/vitality-goes/scripts/Sanchez.sh > /home/sparsons/cronsanchez.log 2>&1

**Reply by JVital2013** [reply]
That's not a valid PATH configuration, AFAIK. The path needs to be the directory that contains executables, not the executables themselves.

You should not need to set the PATH at all, but if you set it, it should probably be something like this:

`PATH=/usr/local/bin:/usr/bin:/bin`

**Reply by parsoli** [reply]
Looks like that fixed it.  It ran at :55.  Now to figure out why it's not creating a timelapse.....must be one of the ini files....Thanks!

sparsons@goesview:~$ cat cronsanchez.log
**[2022-12-23 09:55:02] Creating GOES 16 Sanchez False Color Images
[2022-12-23 09:55:20] Creating GOES16_FD_sanchez_20221223T153021Z.jpg...**
[2022-12-23 09:56:04] Creating GOES 17 Sanchez False Color Images
find: ‘’: No such file or directory
[2022-12-23 09:56:04] Creating GOES 18 Sanchez False Color Images
find: ‘’: No such file or directory
[2022-12-23 09:56:04] Creating GOES 16/17 Composite False Color Images
find: ‘’: No such file or directory
[2022-12-23 09:56:04] Creating GOES 16/18 Composite False Color Images
find: ‘’: No such file or directory

**Reply by JVital2013** [reply]
Looks good!

Make a new top level comment in this discussion topic if you continue to have issues with the video scripts.

**Comment by JVital2013** [comment]
Sanchez.sh will now ignore extra files like .DS_STORE in the source directories as of 6ffee96.
